aten/src/ATen/native/Sorting.cpp

```diff
@@ -271,8 +267,124 @@ std::tuple<Tensor&, Tensor&> kthvalue_out_impl_cpu(
   return std::forward_as_tuple(values, indices);
 }
 
+std::tuple<Tensor&, Tensor&> median_out_impl(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t dim,
+    bool keepdim,
+    bool ignore_nan) {
+  NoNamesGuard guard;
+
+  dim = at::maybe_wrap_dim(dim, self.dim());
+  TORCH_CHECK(self.numel() > 0, "median() input tensor cannot be empty");
+  checkDeviceType("median", {values, indices}, self.device().type());
+  checkScalarType("median", {indices, "indices", 1}, kLong);
+  checkSameType("median", {values, "values", 0}, {self, "self", 2});
+
+  std::vector<int64_t> out_shape = self.sizes().vec();
+  if (self.dim() > 0) {
+    if (keepdim) {
+      out_shape[dim] = 1;
+    } else {
+      out_shape.erase(out_shape.begin() + dim);
+    }
+  }
+
+  resize_output(values, out_shape);
+  resize_output(indices, out_shape);
+
+  Tensor vals = keepdim ? values : values.unsqueeze(dim);
+  Tensor inds = keepdim ? indices : indices.unsqueeze(dim);
+
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "median_out", [&] {
+    dim_apply({self, vals, inds}, dim, [ignore_nan](int64_t it, TensorList tl) {
+      Tensor in = tl[0].contiguous();
```

Comment:
What's up with calling contiguous here, as opposed to calling it once on the input tensor, for example?

Reply:
The idea was that we only need to maintain a copy of the current row to reduce, also, if the dim we're reducing happens to be contiguous (even if it's not the last dim) we don't need to copy at all. Plus, this way multiple threads can perform copies instead of only one at the beginning.

Reply:
A call to contiguous at the start of the function would use multiple threads, right?

Can we compare the perf and memory use of doing this vs. calling contiguous upfront? With like a 1D contiguous, 2D contiguous, and 2D transposed tensor?

Reply:
So what I'm doing now is if the reduction dim is not contiguous, I move it as last dim and copy into contiguous memory once before parallel loop. This is faster especially for single thread.